packages/react/menu/src/MenuTypeahead.tsx

```diff
@@ -0,0 +1,189 @@
+
+const [CollectionSlot, CollectionItemSlot, useCollection] = createCollection<ItemElement, {}>();
+
+/* -----------------------------------------------------------------------------------------------
+ * MenuTypeahead
+ * ---------------------------------------------------------------------------------------------*/
+
+type TypeaheadContextValue = {
+  searching: boolean;
+};
+
+const TYPEAHEAD_NAME = 'MenuTypeahead';
+const TYPEAHEAD_DEFAULT_TAG = 'div';
+
+const [TypeaheadProvider, useTypeaheadProvider] = createContext<TypeaheadContextValue>(
+  TYPEAHEAD_NAME
+);
+
+type MenuTypeAheadPrimitive = Polymorphic.ForwardRefComponent<typeof TYPEAHEAD_DEFAULT_TAG>;
+
+const MenuTypeahead = React.forwardRef((props, forwardedRef) => {
+  const { as = TYPEAHEAD_DEFAULT_TAG, ...typeaheadProps } = props;
+  return (
+    <CollectionSlot>
+      <MenuTypeaheadImpl {...typeaheadProps} as={as} ref={forwardedRef} />
+    </CollectionSlot>
+  );
+}) as MenuTypeAheadPrimitive;
+
+const MenuTypeaheadImpl = React.forwardRef((props, forwardedRef) => {
+  const { getItems } = useCollection();
+  const containerRef = React.useRef<HTMLDivElement>(null);
+  const timerRef = React.useRef(0);
+  const [searchValue, setSearchValue] = React.useState('');
+  const searching = searchValue !== '';
+
+  // Reset `searchValue` 1 second after it was last updated
+  const setSearch = React.useCallback((search: string) => {
+    setSearchValue(search);
+    window.clearTimeout(timerRef.current);
+    timerRef.current = window.setTimeout(() => setSearch(''), 1000);
+  }, []);
+
+  React.useEffect(() => {
+    return () => window.clearTimeout(timerRef.current);
+  }, []);
+
+  React.useEffect(() => {
+    if (searching) {
+      const itemNodes = getItems().map((item) => item.ref.current!);
+      const currentItem = document.activeElement;
+      const currentMatch = currentItem ? getValue(currentItem) : undefined;
+      const values = itemNodes.map((node) => getValue(node));
+      const nextMatch = getNextMatch(values, searchValue, currentMatch);
+      const newItem = itemNodes.find((node) => node.getAttribute(ITEM_ATTR) === nextMatch);
+
+      if (newItem) {
+        /**
+         * Imperative focus during keydown is risky so we prevent React's batching updates
+         * to avoid potential bugs. See: https://github.com/facebook/react/issues/20332
+         */
+        setTimeout(() => (newItem as HTMLElement).focus());
+      }
+    }
+  }, [searchValue, searching]);
+
+  return (
+    <TypeaheadProvider searching={searching}>
+      <Primitive
+        {...props}
+        ref={composeRefs(forwardedRef, containerRef)}
+        onKeyDown={composeEventHandlers(props.onKeyDown, (event) => {
+          const modifierKey = event.ctrlKey || event.altKey || event.metaKey;
+          if (event.key.length === 1 && !modifierKey) {
+            // Perform search on a single menu level only
+            event.stopPropagation();
```

Comment:
I don't love `stopPropagation`, but here it's the more performant option vs triggering n handlers and excluding them.

Reply:
For future reference, I'm thinking we should probably dispatch custom events (that don't bubble) when we want things to `stopPropagation`. It makes me uneasy that we'd be messing with native event behaviour for consumers otherwise.

Reply:
The rub here is that we want to be able to trigger a "typeahead" search when focused on the container as well as the items.

Reply:
We could still rely on bubbling with a custom event and make it cancellable, cause then you're just stopping propagation of a custom event so it's not disrupting native events.

Reply:
oh yeah I see what you both mean. I hadn't thought about it that way, will take a look!